tracer/src/Datadog.Trace/IAST/EvidenceConverter.cs

```diff
@@ -0,0 +1,240 @@
+            if (range.Start > start)
+            {
+                WriteValuePart(writer, value.Substring(start, range.Start - start));
+            }
+
+            string substring = Substring(value, range);
+            WriteValuePart(writer, substring, range);
+            start = range.Start + substring.Length;
+        }
+
+        if (start < value.Length)
+        {
+            WriteValuePart(writer, value.Substring(start));
+        }
+
+        writer.WriteEndArray();
+    }
+
+    private void WriteRedactedEvidence(JsonWriter writer, Evidence evidence, JsonSerializer serializer)
+    {
+        writer.WriteStartObject();
+        if (evidence.Ranges == null && evidence.Sensitive == null)
+        {
+            writer.WritePropertyName("value");
+            writer.WriteValue(evidence.Value);
+        }
+        else
+        {
+            writer.WritePropertyName("valueParts");
+            ToRedactedJson(writer, evidence.Value!, evidence.Ranges, evidence.Sensitive);
+        }
+
+        writer.WriteEndObject();
+    }
+
+    private string Substring(string value, Range range)
+    {
+        if (range.Start >= value.Length) { return string.Empty; }
+        else if (range.Start + range.Length >= value.Length) { return value.Substring(range.Start); }
+        return value.Substring(range.Start, range.Length);
+    }
+
+    private void WriteValuePart(JsonWriter writer, string value)
+    {
+        if (!string.IsNullOrEmpty(value))
+        {
+            WriteValuePart(writer, value, null);
+        }
+    }
+
+    private void WriteValuePart(JsonWriter writer, string value, Range? range)
+    {
+        writer.WriteStartObject();
+        writer.WritePropertyName("value");
+        writer.WriteValue(value);
+        if (range != null && range.Value.Source != null)
+        {
+            writer.WritePropertyName("source");
+            writer.WriteValue(range.Value.Source.GetInternalId());
+        }
+
+        writer.WriteEndObject();
+    }
+
+    private void WriteRedactedValuePart(JsonWriter writer)
+    {
+        writer.WriteStartObject();
+        writer.WritePropertyName("redacted");
+        writer.WriteValue(true);
+        writer.WriteEndObject();
+    }
+
+    private void WriteRedactedValuePart(JsonWriter writer, Range range)
+    {
+        writer.WriteStartObject();
+        writer.WritePropertyName("redacted");
+        writer.WriteValue(true);
+        if (range.Source != null)
+        {
+            writer.WritePropertyName("source");
+            writer.WriteValue(range.Source.GetInternalId());
+        }
+
+        writer.WriteEndObject();
+    }
+
+    private Range? Poll(LinkedList<Range> list)
+    {
+        if (list.First != null)
+        {
+            var res = list.First.Value;
+            list.RemoveFirst();
+            return res;
+        }
+
+        return null;
+    }
+
+    private void ToRedactedJson(JsonWriter writer, string value, Range[]? taintedRanges, Range[]? sensitiveRanges)
```

Comment:
this algorithm is quite complicated. I think some comments giving an overview would be good.